src/hotspot/share/jfr/recorder/service/jfrEvent.hpp

```diff
@@ -146,32 +142,32 @@ class JfrEvent {
     return T::hasStackTrace;
   }
 
-  bool is_started() const {
-    return _started;
+  bool is_started() {
+    return is_instant() || _start_time != 0 || _untimed;
   }
 
   bool should_commit() {
-    if (!_started) {
+    if (!is_enabled()) {
       return false;
     }
     if (_untimed) {
       return true;
     }
-    if (_evaluated) {
-      return _should_commit;
-    }
     _should_commit = evaluate();
     _evaluated = true;
     return _should_commit;
   }
 
  private:
   bool should_write() {
-    return _started && (_evaluated ? _should_commit : evaluate());
+    return _evaluated ? _should_commit : is_enabled_for_thread() && evaluate();
```

Comment:
I wonder if the order for is_enabled_for_thread() and evaluate() should be flipped? Evaluate is more expensive, but 99,999% of all events will be in an enabled thread, so it will be a win on average.

Reply:
I think you have a point there. Will update.

Reply:
Looks good